modules/mobile-agents/android/client/plugins/IDPProxy/src/org/wso2/mdm/agent/proxy/AccessTokenHandler.java

```diff
@@ -0,0 +1,155 @@
+															info.getClientSecret()).getBytes());
+			String encodedCredentials = new String(credentials);
+
+			Map<String, String> headers = new HashMap<String, String>();
+			String authorizationString = Constants.AUTHORIZATION_MODE + encodedCredentials;
+			headers.put(Constants.AUTHORIZATION_HEADER, authorizationString);
+			headers.put(Constants.CONTENT_TYPE_HEADER, Constants.DEFAULT_CONTENT_TYPE);
+
+			Map<String, String> responseParams =
+					ServerUtilities.postDataAPI(apiUtilities, headers);
+			response = responseParams.get(Constants.SERVER_RESPONSE_BODY);
+			responseCode = responseParams.get(Constants.SERVER_RESPONSE_STATUS);
+			return response;
+		}
+
+		@SuppressLint("SimpleDateFormat")
+		@Override
+		protected void onPostExecute(String result) {
+			String refreshToken = null;
+			String accessToken = null;
+			int timeToExpireSecond = 3000;
+			try {
+				IdentityProxy identityProxy = IdentityProxy.getInstance();
+
+				if (responseCode != null && responseCode.equals(Constants.REQUEST_SUCCESSFUL)) {
+					JSONObject response = new JSONObject(result);
+					try {
+						accessToken = response.getString(Constants.ACCESS_TOKEN);
+						refreshToken = response.getString(Constants.REFRESH_TOKEN);
+						timeToExpireSecond = Integer.parseInt(response.getString(Constants.EXPIRE_LABEL));
+						Token token = new Token();
+						token.setDate();
+						token.setRefreshToken(refreshToken);
+						token.setAccessToken(accessToken);
+						token.setExpired(false);
+
+						SharedPreferences mainPref =
+								IdentityProxy.getInstance()
+								             .getContext()
+								             .getSharedPreferences(Constants.APPLICATION_PACKAGE,
+								                                   Context.MODE_PRIVATE);
+						Editor editor = mainPref.edit();
+						editor.putString(Constants.ACCESS_TOKEN, accessToken);
+						editor.putString(Constants.REFRESH_TOKEN, refreshToken);
+						editor.putString(USERNAME_LABEL, info.getUsername());
+						Date date = new Date();
+						long expiresIN = date.getTime() + (timeToExpireSecond * 1000);
+						Date expireDate = new Date(expiresIN);
+						String strDate = dateFormat.format(expireDate);
+						token.setDate(strDate);
+						editor.putString(Constants.DATE_LABEL, strDate);
+						editor.commit();
+
+						identityProxy.receiveAccessToken(responseCode, Constants.SUCCESS_RESPONSE, token);
+					} catch (JSONException e) {
+						Log.e(TAG, "Invalid JSON format.");
+					}
+
+				} else if (responseCode != null) {
+					if (Constants.INTERNAL_SERVER_ERROR.equals(responseCode)) {
+						identityProxy.receiveAccessToken(responseCode, result, null);
+					} else {
+						JSONObject mainObject = new JSONObject(result);
+						String errorDescription = mainObject.getString(Constants.ERROR_DESCRIPTION_LABEL);
+						identityProxy.receiveAccessToken(responseCode, errorDescription, null);
+					}
+				}
+			} catch (JSONException e) {
+				Log.e(TAG, "Invalid JSON." + e);
```

Comment:
Let's make these compressed error messages more consistent across the code. For example, I see some of the error messages carry the exception "cause" appended to the contents where as some don't.